### ISSUE TITLE
DynamicSoftMarginLoss

### ISSUE THREAD

**Issue by ducha-aiki** [KevinMusgrave/pytorch-metric-learning#35]
https://gfx.cs.princeton.edu/pubs/Zhang_2019_LLD/softmargin.pdf

Implemented here, might be possible to borrow with author approval
https://github.com/lg-zhang/dynamic-soft-margin-pytorch/blob/master/modules/dynamic_soft_margin.py#L10
![image](https://user-images.githubusercontent.com/4803565/78721231-6b5a8000-7927-11ea-8407-d3e712ad4b19.png)

**Comment by elanmart** [comment]
Hello @KevinMusgrave , which of the issues tagged with `new algorithm request` do you think would be the most beneficial to have implemented?

I'm looking to contribute to some open source in my spare time and I've always been interested in metric learning.

**Comment by KevinMusgrave** [comment]
Thanks for your interest in contributing!

I think this one (DynamicSoftMarginLoss) or [P2SGrad](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/9#issue-567243436) would be good additions.

Currently, there isn't a lot of documentation for extending the base loss function. And it also became more complicated after the latest version where I added the reducers module. So for now, if you decide to write a loss function, you can refer to these:
- [The base loss function](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/base_metric_loss_function.py).
- [A pair based loss function](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/multi_similarity_loss.py)
- [A classification based loss function](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/normalized_softmax_loss.py)

The function ```compute_loss``` should return a dictionary. The details are a bit complicated. But as an initial working version, you can just compute the loss like you normally do, and then return
```python
return {"loss": {"losses": your_computed_loss, "indices": None, "reduction_type": "already_reduced"}}
```

Let me know if you have any questions!

**Comment by KevinMusgrave** [comment]
The binomial deviance loss would also be a good addition, and would probably be an easier loss to start with. Here's the equation (taken from the [multi-similarity paper](https://arxiv.org/pdf/1904.06627.pdf))
![image](https://user-images.githubusercontent.com/2314891/85235350-3e742000-b3e2-11ea-91eb-3f7f74659867.png)

**Comment by KevinMusgrave** [comment]
Added by @domenicoMuscill0 in v2.4.0:

```
pip install pytorch-metric-learning==2.4.0
```
